[improve][broker,proxy] Use ChannelVoidPromise to avoid useless promise objects creation

nicoloboschi opened the pull request:
### Motivation

In the broker and proxy most of the writes callback are not considered. In those cases we can save useless allocations of default channel promises by using the `voidPromise()` method.

### Modifications

* Use `VoidChannelPromise` if the promise result is never checked

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. Please attach the local preview screenshots (run `sh start.sh` at `pulsar/site2/website`) to your PR description, or else your PR might not get merged. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

nodece commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/LookupProxyHandler.java:
```diff
@@ -446,5 +447,9 @@ private ServerError getServerError(Throwable error) {
     }
 
+    private void writeAndFlushVoidPromise(ByteBuf cmd) {
+        proxyConnection.ctx().writeAndFlush(cmd);
```
```suggestion
        ChannelHandlerContext ctx = proxyConnection.ctx();
        ctx.writeAndFlush(cmd, ctx.voidPromise());
```

nodece commented:
Maybe you should move the `writeAndFlushVoidPromise` method to `pulsar-common` module to eliminate the duplication.

nicoloboschi commented:
@nodece done, PTAL again

lhotari commented:
Related to #16113.

lhotari reviewed (approved):
Really great improvement. The only small nag is about naming. `writeAndFlushVoidPromise` is a very clumsy name. I don't have a good name to suggest immediately, but perhaps we could think about something.

It's not only the "useless promise objects creation" that is the reason to use voidPromise. Another reason to use a `ctx.voidPromise()` is because it adds proper error handling. Exceptions won't be propagated in Netty 4.x unless the the ChannelFuture's listener handles errors or something like `ctx.voidPromise()`.
(Related to SO Q&A https://stackoverflow.com/q/54169262 and https://stackoverflow.com/a/9030420 .)

I think renaming `writeAndFlushVoidPromise` in `PulsarCommandSender` and `ServerCnx` to `writeAndFlush` would simplify naming. Renaming `writeAndFlushVoidPromise` to `writeAndFlushWithVoidPromise` in the utility class would improve the naming.

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarCommandSenderImpl.java:
```diff
@@ -377,7 +375,11 @@ public void sendWatchTopicListUpdate(long watcherId,
+        writeAndFlushVoidPromise(outBuf);
+    }
+
+    private void writeAndFlushVoidPromise(ByteBuf outBuf) {
```
```suggestion
    private void writeAndFlush(ByteBuf outBuf) {
```

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarCommandSenderImpl.java:
```diff
@@ -377,7 +375,11 @@ public void sendWatchTopicListUpdate(long watcherId,
+    }
+
+    private void writeAndFlushVoidPromise(ByteBuf outBuf) {
+        NettyChannelUtil.writeAndFlushVoidPromise(cnx.ctx(), outBuf);
```
```suggestion
        NettyChannelUtil.writeAndFlushWithVoidPromise(cnx.ctx(), outBuf);
```

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/protocol/PulsarDecoder.java:
```diff
@@ -731,4 +739,8 @@ protected void handleCommandWatchTopicListClose(CommandWatchTopicListClose comma
 
     private static final Logger log = LoggerFactory.getLogger(PulsarDecoder.class);
+
+    private void writeAndFlushVoidPromise(ChannelOutboundInvoker ctx, ByteBuf cmd) {
```
```suggestion
    private void writeAndFlush(ChannelOutboundInvoker ctx, ByteBuf cmd) {
```

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/protocol/PulsarDecoder.java:
```diff
@@ -731,4 +739,8 @@ protected void handleCommandWatchTopicListClose(CommandWatchTopicListClose comma
     private static final Logger log = LoggerFactory.getLogger(PulsarDecoder.class);
+
+    private void writeAndFlushVoidPromise(ChannelOutboundInvoker ctx, ByteBuf cmd) {
+        NettyChannelUtil.writeAndFlushVoidPromise(ctx, cmd);
```
```suggestion
        NettyChannelUtil.writeAndFlushWithVoidPromise(ctx, cmd);
```

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/netty/NettyChannelUtil.java:
```diff
@@ -0,0 +1,32 @@
+ */
+public class NettyChannelUtil {
+
+    public static void writeAndFlushVoidPromise(ChannelOutboundInvoker ctx, ByteBuf cmd) {
```
```suggestion
    public static void writeAndFlushWithVoidPromise(ChannelOutboundInvoker ctx, ByteBuf cmd) {
```

lhotari commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/LookupProxyHandler.java:
```diff
@@ -446,5 +448,10 @@ private ServerError getServerError(Throwable error) {
         return responseError;
     }
 
+    private void writeAndFlushVoidPromise(ByteBuf cmd) {
```
```suggestion
    private void writeAndFlush(ByteBuf cmd) {
```

lhotari commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/LookupProxyHandler.java:
```diff
@@ -446,5 +448,10 @@ private ServerError getServerError(Throwable error) {
 
+    private void writeAndFlushVoidPromise(ByteBuf cmd) {
+        final ChannelHandlerContext ctx = proxyConnection.ctx();
+        NettyChannelUtil.writeAndFlushVoidPromise(ctx, cmd);
```
```suggestion
        NettyChannelUtil.writeAndFlushWithVoidPromise(ctx, cmd);
```

lhotari reviewed (changes requested):
Please also replace any occurrences of  `.addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE)` with the use of `ctx.voidPromise()` solution. I realised that I had added `.addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE)` in #14713 based on SO answer https://stackoverflow.com/a/54377238 . (I added [a new answer to SO with the voidPromise solution](https://stackoverflow.com/a/75066368).)

nicoloboschi commented:
Thanks @lhotari 
I've renamed the methods and javadoc as you suggested. I've also replaced the `.addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE)` usages and added unit tests. 

Lastly, I've also added a convenience method for closing the channel after the write.
PTAL again

lhotari reviewed (approved):
Good work @nicoloboschi

lhotari commented:
/pulsarbot rerun-failure-checks

nodece commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java:
```diff
@@ -495,5 +493,9 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
     }
 
+    private void writeAndFlush(ByteBuf cmd) {
+        NettyChannelUtil.writeAndFlushWithVoidPromise(outboundChannel, cmd);
```
Do you miss `FIRE_EXCEPTION_ON_FAILURE`?

lhotari replied:
@nodece VoidChannelPromise [will propagate exceptions](https://github.com/netty/netty/blob/0097c2b72521c422d42bf8c45273db05a7ff068c/transport/src/main/java/io/netty/channel/VoidChannelPromise.java#L177-L180).

michaeljmarshall reviewed (approved):
LTMG. Great catch!
